docs/workforce-human-oidc-auth.md

```diff
@@ -0,0 +1,267 @@
+behave accordingly.
+
+Since these steps require the involvement of a human being, the application should be set up to gracefully handle
+timeouts and provide the user with a way to abort the connection attempt.
+
+# General requirements
+
+An application implementing workforce OIDC needs to generally comply with OpenID and OAuth 2.0 standards.
+
+In particular:
+
+- Any HTTP calls made to non-local servers MUST be made using HTTPS.
+- IdP metadata discovery MUST be implemented as described in [RFC8414](https://datatracker.ietf.org/doc/html/rfc8414).
+
+Using a well-tested and standards-compliant third-party library for core OIDC logic for the respective ecosystem is
+highly recommended. If this is not possible, implementers need to pay special attention to the specifications referenced
+in this document.
+
+## Endpoint restrictions
+
+Currently, users who connect to a host other than localhost or an Atlas hostname need to explicitly opt-in into being
+able to do so by setting the `ALLOWED_HOSTS` flag (specified in the drivers auth spec). In the future, MongoDB is hoping
+to support Demonstrating Proof of Possession (DPoP, [RFC9449](https://datatracker.ietf.org/doc/html/rfc9449)) that will
+allow lifting this restriction. The goal here is to prevent users from connecting to untrusted endpoints that will
+advertise attacker-controlled IdP metadata and intercept tokens intended for other clusters (or even other OIDC
+endpoints in general).
+
+We would also like to generally adopt [RFC8707](https://datatracker.ietf.org/doc/html/rfc8707), but have not decided on
+a specific format for expressing the MongoDB endpoints as resources.
+
+## Token management
+
+After a successful authentication, applications SHOULD periodically attempt to use the OIDC token refresh mechanism in
+order to exchange the access token it uses for a fresh one. The refresh interval SHOULD be determined by the token
+expiry time indicated by the Identity Provider. If the refresh attempt fails, the application MAY ask or suggest the
+user to re-authenticate from scratch, even if the currently used access token has not expired yet.
+
+Under normal circumstances, the application provides an access token received from the Identity Provider to the driver.
+The application SHOULD provide a toggle to users that can be used to indicate that it should use the ID token, rather
+than the access token, in its place.[^2]
+
+## OIDC Scopes
+
+The initial request to the IdP involves specifying the `scope` parameter. This is a space-separated unordered list of
+strings which contains:
+
+- The scopes listed in the `requestScopes` field of the IdP response from the MongoDB endpoint, and
+- The `openid` and `offline_access` scopes[^3]. If the IdP metadata document contains a `scopes_supported` field, and
+    this field does not list one of these scopes, then the respective scope SHOULD NOT be added to the `scope`
```

Comment:
Is this a SHOULD NOT or a MUST NOT? Does auth fail if an unsupported scope is provided?

Reply:
Hm, yeah, it's a good question, and I've gone a bit back and forth on this in response to you bringing it up 🙂 

I think the RFC2119 definition of "SHOULD NOT" is ultimately the right one here. The implementer would need to accept that their application's authentication attempts could be rejected by an IdP that doesn't provide refresh tokens (as that is what `offline_access` asks for). That seems like a sensible behavior for an application that strictly requires ongoing access to the cluster while it is active – e.g. mongosync might want to make this choice (Not saying they do, but that it would be understandable if they did).

That being said, if you feel strongly here, I'm still happy to make this change.

> Does auth fail if an unsupported scope is provided?

It's up to the individual IdP (this sentence referring to the IdP/server side of OIDC):

> Scope values used that are not understood by an implementation SHOULD be ignored.